WIP: Read debug symbols provided by the DDC

annagrin opened the pull request:
This is a first in a series of changes to support dwds translating JS objects to dart when displaying the objects to the user, for example, in variable inspection, watch, or expression evaluation windows.

The aim is to make the translation guess-free and reduce the number of calls to chrome. 

The symbols are currently stored by DDC in https://dart-review.googlesource.com/c/sdk/+/200528.

annagrin commented on dwds/lib/src/debugging/metadata/module_symbols.dart:
```diff
@@ -0,0 +1,758 @@
+// Dart debug symbol information stored by DDC.
+//
+// The data format below stores descriptions of dart code objects and their
+// mapping to JS that is generated by DDC. Every field, except ids, descrbes
```
thanks!

annagrin commented on dwds/lib/src/debugging/metadata/module_symbols.dart:
```diff
@@ -0,0 +1,758 @@
+
+// Note that 'localId' and 'scopeId' combine into a unique id that is used for
+// object lookup and mapping between JS and dart concepts. As a result, it
+// needs to be either stored or easly computed for each corresponding JS object
```
thanks!

annagrin commented on dwds/lib/src/debugging/metadata/module_symbols.dart:
```diff
@@ -0,0 +1,758 @@
+// needs to be either stored or easly computed for each corresponding JS object
+// created by DDC, so the debugger is able to look up dart symbol from JS ones.
+//
+// For example, to detect all dart variables in current scope and display
```
thanks!

annagrin commented on dwds/lib/src/debugging/metadata/module_symbols.dart:
```diff
@@ -0,0 +1,758 @@
+  }
+}
+
+abstract class SymbolTableElement {
```
done

annagrin commented on dwds/lib/src/debugging/metadata/module_symbols.dart:
```diff
@@ -0,0 +1,758 @@
+  }
+
+  @override
+  Map<String, dynamic> toJson() {
```
I will leave to post-prototype phase.

annagrin commented on dwds/lib/src/debugging/metadata/symbol_provider.dart:
```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
```
thanks!

annagrin commented on dwds/lib/src/debugging/metadata/symbol_provider.dart:
```diff
@@ -0,0 +1,118 @@
+  final Map<String, ScopeSymbol> scopes = {};
+  final Map<String, VariableSymbol> variables = {};
+
+  SymbolTable();
```
I am adding some functionality to the constructor, so it will be soon.

nshahan commented on dwds/test/fixtures/context.dart:
```diff
@@ -172,6 +174,10 @@ class TestContext {
               ],
+              if (enableDebugSymbols) ...[
+                '--define',
+                'build_web_compilers|ddc=generate-debug-symbols=true',
```
nit: Is this name consistent with other build_web_compiler options for DDC? Do we need the word "generate" in the option or would `debug-symbols=true` be enough?

annagrin replied:
We only have another one and it is used above: `generate-full-dill` so I prefer them to be similar. I am open to suggestions to what to rename them  - I can change them at the same time in a series of follow-up changes.

annagrin replied:
Might rename to match the sdk...

annagrin replied:
Renamed to `output-debug-symbols` to be consistent with the SDK and build_web_compilers options. Will cleanup existing options in a separate PR.

nshahan commented on webdev/lib/src/command/shared.dart:
```diff
@@ -60,6 +60,11 @@ void addSharedArgs(ArgParser argParser,
+        abbr: 's',
+        defaultsTo: false,
+        negatable: true,
+        help: 'Enable expression evaluation features in the debugger.')
```
Is this meant to have the exact same description as --enable-expression-evaluation? Will we eventually remove this special flag before publishing the stable version of webdev 2.8.0?

annagrin replied:
> Is this meant to have the exact same description as --enable-expression-evaluation?

No, my mistake. Thanks, fixed!

> Will we eventually remove this special flag before publishing the stable version of webdev 2.8.0?

I plan to remove each feature flag when the feature is stable and is enabled by default in webdev, google3, and flutter tools. We are close with expression evaluation but not exactly there - it is still disabled by default in webdev.
Variable inspection using symbols will need some time to get there though.

grouma commented on dwds/lib/src/debugging/metadata/symbol_provider.dart:
```diff
@@ -0,0 +1,485 @@
+  final Map<String, VariableSymbol> variables = {};
+  final Map<String, FunctionSymbol> globalFunctions = {};
+
+  // build-in types
```
[nit] built-in

annagrin replied:
removed those comments, thanks!

grouma commented on webdev/lib/src/command/shared.dart:
```diff
@@ -60,6 +60,11 @@ void addSharedArgs(ArgParser argParser,
         defaultsTo: false,
         negatable: true,
         help: 'Enable expression evaluation features in the debugger.')
+    ..addFlag(enableDebugSymbolsFlag,
```
[optional] we could keep this flag hidden

annagrin replied:
Will do!

annagrin replied:
Done, thanks!

grouma commented on dwds/lib/src/debugging/metadata/symbol_provider.dart:
```diff
@@ -0,0 +1,485 @@
+    for (var scope in scopeChain) {
+      for (var id in scope.variableIds) {
+        var variable = getVariableSymbol(id);
+        if (variable.location.endTokenPos < tokenPos) {
```
Can you elaborate on this. I don't follow.

annagrin replied:
This code collects all dart variables in scope at the current position in the source (tokenPos). 
Given the current scope represented by `symbol`, and a location in it (`tokenPos`), we walk the scope chain starting from the library and ending by the innermost scope, collecting all variables in scope (ie defined in the scope chain above the tokenPos.

Added comment.

grouma replied:
I'm more confused on how the tokenPos are defined. Things in an outer scope defined past a tokenPos should be considered in scope eh?

annagrin replied:
Not for local and formal variables, the ones defined past the current position should not be in scope at the current position. But I definitely have a bug here, as global variables in library and  fields in class are stored as variableIds as well... will fix, thanks!

annagrin replied:
Fixed.

grouma commented on dwds/lib/src/debugging/metadata/symbol_provider.dart:
```diff
@@ -0,0 +1,485 @@
+    return variables;
+  }
+
+  void addSymbols(ModuleSymbols symbols) {
```
Does this need to be public?

grouma commented on dwds/lib/src/debugging/metadata/symbol_provider.dart:
```diff
@@ -0,0 +1,485 @@
+  }
+
+  void validate() {
+    for (var e in scopes.values) {
```
Where is this called? Does this need to be public?

nshahan commented on dwds/lib/src/debugging/metadata/provider.dart:
```diff
@@ -65,7 +65,8 @@ class MetadataProvider {
         'dart:web_gl',
-        'dart:web_sql'
+        'dart:web_sql',
+        'dart:ui',
```
Web projects that are not flutter web will not have this library in the SDK. Is the code that depends on this list flexible enough to handle that?

annagrin replied:
This is used only to create a VM  Service protocol `Library` and `LibaryRef` objects on requests from the user of dwds (a UI), and those requests would come from evaluating objects from those libraries, so there is no problem in cases you mentioned.

nshahan reviewed (approved):
Left one more question for you but LGTM

annagrin commented:
Updated build_web_compilers now that  https://github.com/dart-lang/build/pull/3120 is published.
Updated min SDK constraint to 2.14.0-216.0.dev to match build_web_compilers constraints.

annagrin commented:
@grouma I updated min SDK and build_web_compilers constraints, could you please have another look?
